### DISCUSSION TITLE
SecRandomCopyBytes in Sign with Apple

### DISCUSSION THREAD

**Discussion by jesus-mg-ios** [firebase/firebase-ios-sdk#10732] [Q&A] [answered]
Hi:

I am reviewing sign with apple, and I notice that SecRandomCopyBytes is called multiple times (1 per byte) instead of fill the nonce bytes length at once. Is this related to something about the strength of the nonce?

Thanks in advance for the clarification

Extracted from https://firebase.google.com/docs/auth/ios/apple :

```Swift
let randoms: [UInt8] = (0 ..< 16).map { _ in
      var random: UInt8 = 0
      let errorCode = SecRandomCopyBytes(kSecRandomDefault, 1, &random)
      if errorCode != errSecSuccess {
        fatalError(
          "Unable to generate nonce. SecRandomCopyBytes failed with OSStatus \(errorCode)"
        )
      }
      return random
    }
```

What I think that can be done

```Swift
 let length = 32
 let randomBytes = [Int8](repeating: 0, count: length)
 let errorCode = SecRandomCopyBytes(kSecRandomDefault, length, &random) 
 if errorCode != errSecSuccess {
        fatalError(
          "Unable to generate nonce. SecRandomCopyBytes failed with OSStatus \(errorCode)"
        )
 }
```

**Comment by ncooke3** [comment, marked as answer]
Hi @jesus-mg-ios, thanks for pointing that out. I can't find any documentation that specifies that calling `SecRandomCopyBytes` multiple times (1 per byte) instead of once (with multiple bytes) results in nonces with different cryptographic strengths.

In your simplified code snippet, I think the `length` should be `16` to match that of the original snippet's number of bytes, right?

**Reply by jesus-mg-ios** [reply]
Right, if I'm not wrong the value of the length by default is 32 chars/bytes, so we can do it with one loop iteration in a happy path .

Also I think that the rest of code could be improved, at least in legibility for last swift versions, I left my code snippet if it's interesting :

```Swift
// length is the arg passed to the function 
let charset =  Array("0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._")
var result = ""

while(result.count < length) {
    // We fill all random bytes at once
    let randomBytes = [Int8](repeating: 0, count: length) // Length of the nonce required 
    let errorCode = SecRandomCopyBytes(kSecRandomDefault, length, &random) // fill the length bytes
    if errorCode != errSecSuccess {
       assertionFailure("Unable to generate nonce. SecRandomCopyBytes failed with OSStatus \(errorCode)") // Not use fatal error because it will crash on production
       continue // If something was wrong, loop again
   }

    for randomByte in randomBytes where Int(randomByte) < charset.count {
        result.append(charset[Int(randomByte)]) // Filling the result of nonce
    }
}
```

**Reply by ncooke3** [reply]
Thanks @jesus-mg-ios, I agree that the sample could be improved and will take a look!

**Reply by jesus-mg-ios** [reply]
@ncooke3, is "review the sample" on the backlog?

**Reply by ncooke3** [reply]
yes, sorry @jesus-mg-ios, thanks for the ping. I will see about getting this updated this upcoming week.

**Reply by jesus-mg-ios** [reply]
Nice, thanks for your hard work!

**Reply by ncooke3** [reply]
I played around with it again and was trying to figure out if there was a good way to avoid the `while` loop. I came up with the below snippet. I _think_ this should be cryptographically equivalent with the current snippet hosted on the docs. I chose to throw an error for the failure case just to see what that looks like.

```swift
enum NonceGenerationError: Error {
    case generationFailure(status: OSStatus)
}

private func randomNonce(length: Int = 32) throws -> String {
    var randomBytes = [UInt8](repeating: 0, count: length)
    let errorCode = SecRandomCopyBytes(kSecRandomDefault, randomBytes.count, &randomBytes)
    if errorCode != errSecSuccess {
        throw NonceGenerationError.generationFailure(status: errorCode)
    }

    let charset: [Character] =
      Array("0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._")

    let nonce = randomBytes.map { byte in
        // Pick a random character from the set, wrapping around if needed.
        return charset[Int(byte) % charset.count]
    }

    return String(nonce)
}
```

**Reply by jesus-mg-ios** [reply]
Yas! LGTM
